File: solutions/c/binary-search/1/binary_search.c
#include "binary_search.h"

const int *binary_search(int value, const int *arr, size_t length)
{
    int *result = (int *)arr;
    
    while (length > 0)
    {
        if (result[length / 2] == value)
            return (&(result[length / 2]));
        else if (result[length / 2] > value)
        {
            length /= 2;
        }
        else
        {
            length /= 2;
            result += length + 1;
        }
    }
    return (0);
}

File: solutions/fortran/binary-search/2/binary_search.f90
module binary_search
  implicit none

contains

  function find(array, val) result(idx)
    integer, dimension(:), intent(in) :: array
    integer, intent(in) :: val
    integer :: idx
    integer :: low, high, mid

    low = 1
    high = size(array)
    idx = -1

    do while (low <= high)
      mid = (low + high) / 2
      if (array(mid) == val) then
        idx = mid
        return
      elseif (array(mid) < val) then
        low = mid + 1
      else
        high = mid - 1
      end if
    end do

  end function

end module binary_search
